### PR TITLE
Drill-Through + Actions

### PR DESCRIPTION
**Pull request by tlrobinson**
Sherpa: @salsakran 

### Feature To-dos
- [ ] Replace Summarize for Raw Tables with 
   - common metrics (flattened, and limited to 5?)
   - count of TableName(plural) by time
   - sum of columns -> column picker???
- [x] add drill through options to the chart legends for a multi-series chart. These should be the same as result cell drill through, but keeping the filter for that series. (Stretch) 

### Bugs
- [x] When clicking on a raw data cell, the “View these TABLENAME” entry should not be visible
- [x] When clicking on a scalar, a list of times appears. If that is the only option, it should still have a top level entry
- [x] Raw tables don't seem to have drill through enabled any more
- [x] If I click on "View the underlying records", it takes 2 back button clicks to get back to the starting query

### Polish
- [x] Scalars drill through should have some kind of hover state
  - Besides `cursor: pointer`? -tom ---> YES, it's way too subtle
- [x] Seeing both “view this User” and “view these Users” on a time series drill through top level menu
- [ ] When clicking on a funnel, the target should be the line and not the bar
- [x] When drilling into a Quarter, the filter is set to a single date (July 1 2017) instead of the quarter (Q3 2017). The results look to be correct.
- [x] If a query already includes a breakout by a dimension, it should not be in the list of available dimensions to pivot by on drill through or actions
- [x] Mode should only be updated after a query returns .. i.e. “keyed off result set, not query dictionary”

### Open-ended To-dos
- [ ] Decide on ordering of options 
- [ ] View these X” should be renamed “View rows where ColumnName=Value?” or something else?

### Open Questions
- Should we ship the time series footer as part of this?
- Can we make it more clear what the footer does? Currently there are no labels, and the addition of filters next to granularity makes what was simple (the connection of granularity to the time axis) confusing.
- How does this work in the context of permissions?
- How does this work wrt public links?

### CODE REVIEW

**Comment by mazameli** [comment]
@tlrobinson and I synced up today on this, and discussed the immediate next steps as being:
- Get SQL mode in there
- Do some basic wiring up of the new `new question` flow so that SQL mode and tables/segments are selectable
- Get a `get answer` / `refresh` button in so queries can be edited and refreshed

Some of the big questions are around how SQL works, and how visualization settings work. One thing I noted in the list in the description for this issue is that it's currently unclear how to go from an unaggregated segment/table to a visualization, because the chart settings as currently implemented don't include a direct visualization type choice. Do we want to tackle part of or all of the _new_ visualization settings design?

**Comment by salsakran** [comment]
How much of the charting selection can happen as actions? eg "map this segment" sort of stuff

**Comment by mazameli** [comment]
Hmm, I guess it depends on how many actions we'd want to list, since unaggregated tables can be charted with pretty much any/every visualization. Maybe you're talking about just listing actions for the visualization choices that wouldn't require any additional user input? Like "map this segment" could show a pin map if the table has lat+long without requiring additional choices; or viewing a single metric on a region map, you could have an action that says "view over time" (though that's what I thought the "pivot by time" action was supposed to do).

**Comment by salsakran** [comment]
one thing that's weird is that when I am in timeseries mode and select a point and click on "view orders" and hit the browser back button, I don't return to the timeseries view. It takes an additional back button press.

**Comment by salsakran** [comment]
Also in time series mode, when I pivot the a given point by a category, the breakout clause is added but the visualization is remains the same.

**Comment by salsakran** [comment]
in general, when a 1-d pivot is pivot-ed again, we should switch to a table view unless the chart settings allow 2D1M results

**Comment by salsakran** [comment]
the new question flow for tables (while ugly) sorta kinda works. It generates a valid url, and if I hit refresh afterwards it works, but stalls out on the last step.

**Comment by salsakran** [comment]
when clicking on a map cell, I only see a "view list" option.

screen shot 2017-03-22 at 4 53 06 pm

**Comment by salsakran** [comment]
pivot table cells should also have the same drill through options they would have in a chart

**Comment by mazameli** [comment]
Here's a bunch of collected thoughts. I'll go through this and put the clear-cut implementation-related ones into the to-do momentarily. Many of the others probably require discussion.

**Action bauble button**
* Explore bauble is super subtle. As discussed in Slack, we’ll need to get this in front of some people and see if they notice it.
* I think actions in the menu should be disabled if there aren’t any usable fields. (E.g., if there aren’t any time fields, “pivot by time” should either be disabled or not present). And if there are multiple tables listed (e.g. the source table and FK tables), if a table doesn’t have any relevant items, I don’t think we should show the heading label for that table.
* If I pivot by category, and my category list is super long, I can’t scroll through it to see all the items.
* The difference between “View the underlying data” and “View the underlying [table name] records” gives me some cognitive load. I still think there’s merit to the idea of pulling out “view the underlying data” into its own distinct, persistent toggle. Like a “Chart (   O) Data” switch. Otherwise it’s unclear how to get back to your chart, and modifying the query after switching over to Table view takes away the possibility of using the browser’s back button.
* A few times, I found myself instinctively going to the bauble to refresh the query. Not sure if there’s something to that, e.g. moving the Get Answer/Refresh button into the menu, but it felt worth noting.
* Pivoting a scalar by time and getting a time series chart as a result, or by state and getting a map as a result, feels really great. It’s making me sometimes wish that, when I modify a query’s breakouts directly via the QB, that I would get the best-guess visualization automatically.
* I wouldn’t mind a tasteful, subtle drop shadow on this good little boy.

**Time series footer**
* Defaulting to the “Between” calendar felt pretty overwhelming to me.
* The popover can get super tall. The way it resizes overall feels pretty janky right now.
* If I didn’t know what the granularity picker was, I don’t think I’d know what it does. Like I land on a question broken out by month, and at the bottom of the screen I just see two boxes that say “All Time” “Month.”
* The granularity picker could really use some grouping/chunking/simplifying. 13 options feels crazy. If it weren’t for the bottom 6 options (the x of y ones), I’d almost say we could use a slider within the popup for the granularity.

**Drill throughs**
* Popover still needs design love. At minimum I’d suggest removing the button boxes and left-aligning the action text.
* The logic the drill-through menu uses to decide what noun/subject to use for each action is kind of confusing. If you have count of rows of Orders by Year, clicking on a dot on the line will give you, “View this year,” but also “View these Orders.” We might’ve gotten too specific/cute.
* For series with two breakouts, it might be neat to add drill through to chart legends. Like, “looks like Doohickeys are selling the best. let me click on that category legend to just see Doohickey orders,” which would in this case remove the Category breakout and add “Is Doohickey” as a filter.
* If I have the count of orders by state as a bar graph and I click on the California bar, beyond having the option to see those rows it would be neat to “see these orders over time,” as a shortcut to change the breakout from State to Created At, and filter by State=California.

**New Question flow**
* Where and how will subqueries (i.e. picking a saved question as a starting table) fit into this?
* Map screen is goofy to me. Map itself draws too much attention, and it’s hard to tell what I’m supposed to click.
* On screens that include a list of existing metrics, how are we going to group this list, or filter/sort it, if there are dozens of metrics?
* Links in “existing metrics” list don’t have `cursor: pointer`
* If I pick “metric on a map,” and then pick a metric that doesn’t have a Location field, I still see all my Number and Category fields. I would expect that, probably, I shouldn’t even see metrics listed after picking “metric on a map” if there isn’t an available Location field. Either that, or let me pick that metric, but then show me a state that says, “hmm we didn’t find a location field. Want to try picking one?” to contextualize the fact that the field I pick might not result in a map.
* “Pivot a metric” section seems to not do anything after clicking on a metric in the list. This sometimes happens in scalar mode (“a metric”), too; sometimes requires a hard refresh to get the QB page to load the metric.

**Comment by mazameli** [comment]
This is probably a dumb idea, since ultimately I think we're going to want to get rid of the bauble button and move the actions into the future sidebar QB, but what if we made the Run Query / Refresh button a circle and put it next to or above the explore bauble? Might make the explore bauble more discoverable, and might even feel like a nicer home for the run button. ¯\_(ツ)_/¯

**Comment by salsakran** [comment]
Seeing a few bugs
1) when I click on a cell in a raw table view, eg a date, I get a `view these <TABLENAME>` option
screen shot 2017-03-27 at 2 59 48 pm
2) when I click on a scalar, I get a list of times (assuming this is a short circuiting of available options)
screen shot 2017-03-27 at 3 00 34 pm
3) on time series seeing both `view this TABLENAME` and `view these TABLENAMES`
4) timeseries mode time filter doesn't work. clicking on `apply` below doesn't do anything.
screen shot 2017-03-27 at 3 02 52 pm
JS console logs
screen shot 2017-03-27 at 3 03 40 pm
5) clicking on a cell in a pivot table should provide the same drill-through options you get when clicking on that cell in a chart (eg, when you click on that bar)

**Comment by salsakran** [comment]
I'm not sure it makes sense to provide a "filter by this cell result" option. Eg, I'm looking at a state heatmap, and I click on a state. If I filter by that state, all it does is remove all the other states. 

I get that filtering might be useful, but not sure this specific method is worth including in the v1 list.

**Comment by mazameli** [comment]
Incoming brain dump!

* It feels pretty odd to not have the drillthrough actions on dashboard cards once you get used to them on individual questions.
* I don’t think drill-through for series with breakouts is working right. If I have a bar chart with count of orders broken out by time and product category, if I drill through on a single bar to see the unaggregated rows, I’d expect to have those rows be filtered by _both_ time _and_ category. Currently it appears to only be applying the time filter.
* Probably not a regression but when you drill through on a quarter, the filter value that’s shown in the query builder gives a specific date, like July 1 2017 (the first day of the quarter), instead of something like Q3, 2017.
* I feel more sure now that a dedicated chart/data toggle would be great.
* When clicking on a bar in a bar chart, I don’t know what “Filter by this value” means — the value of the x-axis or the y-axis?
* Actions don’t yet work on funnel segments
* When trying to summarize by a metric in a segment, after selecting the aggregation type, when I try to use the field search box, after I start typing I get bounced back to the list of aggregation types.
* If I do a new query with an aggregation broken out by time, I expect to see a line chart time series, but I’m seeing a table instead.
* The time series chrome filters (range and granularity) don’t seem to be working: they open, but I can’t select a value. But then if I manually refresh the query (with the Refresh button) it then _does_ apply the filter or granularity change.
* I don’t know if we’ve chosen an intentional or logical grouping/ordering for the actions you see when clicking on a chart. When there are around 5 possible actions, the list is hard to parse.
* The pathway you can go via actions feels so cool: from Orders by Quarter, to Orders during this quarter by product category, to the list of specific orders for the best selling product, to the average subtotal of those orders, to seeing those orders on a map — very cool.
* Tiny thing: sometimes a drill-through action, like “View these Orders”, only takes you to a single record. It would be an extra credit point if it said “View *this* Order”.
* “View this user” from a table cell with a user ID feels cool.
* I know this isn’t a regression, but I _completely_ missed the action you can take from a single user’s record to view all their orders. (The connected-to side bar.) I was going to write that I wished I could do exactly that. Maybe we should include that action in the explore bauble when on a single record view.
* When viewing a single user, the explore actions feel cold and heartless: “summarize by a metric in this segment” — seems weird when I’m looking at a single person. It also lets you choose “plot a field in this segment” for a single record, which doesn’t do anything currently (not really sure what it could ever do). Weirdly, if you click on that plot action, the menu of actions changes: it then includes “view the underlying data” though the visualization didn’t change.

**Comment by mazameli** [comment]
The new language for the action to see the unaggregated rows for a single "cell" really threw me off.  E.g., `Filter Orders by this value` is super abstract to me compared to `See the list of these Orders` or something similar. The key thing that's missing for me is the `list` word; I did not at all get that this was going to show me the unaggregated rows.

**Comment by mazameli** [comment]
This is outside the scope of this PR, but it would be neat if, when changing the zoom level of a time series, we could do an animated transition like [the example on this page](https://developers.google.com/chart/interactive/docs/animation) under the "Changing the view window" section. (In our case the x-axis would not only move, but of course the axis labels themselves would also need to change.) I think it might help keep the user oriented as to where (or when?) they are.

**Comment by mazameli** [comment]
> Scalars drill through should have some kind of hover state

I think the simplest thing we could do is `color: #27649A` and `cursor: pointer`, ideally with a real quick (~0.3s) CSS transition on `color`. How about we just start there and see if we need more.

**Comment by mazameli** [comment]
> "View these X” should be renamed “View rows where ColumnName=Value?” or something else?

I don't mind the suggestion above, though I'd change `=` to `is`. If `columnName` is a fk field, I suppose we could style it something like `Product:Category`. (It's tempting to get rid of the foreign table identifier, but then it's impossible to distinguish between `Orders:Created At` and `Product:Created At`.)

Initially I'd wanted to display the table name, like `View the list of [table name] where [field name] is [value]`, but that could get potentially too lengthy. An example worst case might be, 
`View the list of Catalog Category Anc Cate Gs Index I Dx where Value is 10`

**Comment by mazameli** [comment]
> Number itself feels kind of small and maybe oddly placed

This was more the case when the branch included the new sidebar. I don't think it's the case anymore with the current state of the branch.

> Once I'm visualizing a segment as a chart, how do I get out of that visualization mode?
> Going from viewing the underlying data (for say a time series) back to the chart feels weird
> Explore Data <-> Chart Toggle in place of "view underlying data"

These are all related, and after talking with Sameer yesterday my current thinking is that we'll want to consider the data/chart toggle in the context of a slightly larger revisiting of charting actions and UI, which might necessitate the sidebar. I'm not opposed to trying out a toggle next to the current visualization type dropdown for now, but I suspect it might feel a little gross.

**Comment by mazameli** [comment]
> Can we make it more clear what the footer does? Currently there are no labels, and the addition of filters next to granularity makes what was simple (the connection of granularity to the time axis) confusing.

I'd like to try two subtle things: adding a little calendar icon to the front of the filter widget, and just adding the word `by` to the granularity picker, like this:

granularity

This would result in combinations that would read like `Today` `by hour`, or `March 1, 2016 - April 15, 2016` `by day`.

**Comment by salsakran** [comment]
@mazameli I'm not sure when it happened, but whatever decision lead to the two selects being next to each other is causing a lot of confusion.

The "by X" should be connected to the name of the dimension, eg "created_at". Grouping it with the filter is non-sensical. I'd rather fix the upstream (in IA terms) issue than add bandaids.

**Comment by mazameli** [comment]
My personal impression was that the filters themselves were unclear due to lack of labeling, not necessarily due to proximity to each other. But I hear where you're coming from, and I'm happy to step back from this and return to the overall IA problem.

As far as when it happened, I recall a design review when Kyle walked you and me through a number of possibilities and we landed on the current arrangement as the best of the presented options.

**Comment by mazameli** [comment]
Not sure if we punted on the below or if they slipped through the cracks (not seeing these in the description list):

- The time filter's list of types (`Previous`, `Current`, `Between` etc.) is too tall and doesn't let you scroll through all of the available options. You can't get to `Previous` currently, for example.
- Related to that, I would far prefer if, when opening the filter popup, we defaulted to `Previous 30 Days` like we do with the filter in the query builder. It's way less imposing than the massive `Between` filter.

**Comment by tlrobinson** [comment]
> The time filter's list of types (Previous, Current, Between etc.) is too tall and doesn't let you scroll through all of the available options. You can't get to Previous currently, for example.

This is also true of the existing date filter widget, isn't it?

**Comment by mazameli** [comment]
> This is also true of the existing date filter widget, isn't it?

The difference is that the existing one still works since you can scroll down the page. Because the new filter is at the bottom of the page, when you open the popover it gets clipped off the top of the screen.

**Comment by mazameli** [comment]
Drill through weirdness: the action labels for drill throughs on time series are using the wrong time granularity language (they're off by one level). E.g., if I click on a bar in a bar chart that's broken out by day, it says `Drill into this hour`, but it should say `Drill into this day` since each bar represents a day.

**Comment by mazameli** [comment]
Possible bug: when looking at a question with a visualization other than `table`, if I then change the source table the visualization of the _current_ query becomes a table before I've done anything else. It's a bit jarring and momentarily makes you think you're looking at the _new_ table's raw data.

**Comment by salsakran** [comment]
"If I click on "View the underlying records", it takes 2 back button clicks to get back to the starting query" is still happening as of the latest commit for me.

**Comment by tlrobinson** [comment]
> "If I click on "View the underlying records", it takes 2 back button clicks to get back to the starting query" is still happening as of the latest commit for me.

@salsakran Can you give me exact steps to repro? I'm not seeing this.
